### ISSUE TITLE
No way to use alternate metrics, like Mahalonobis

### ISSUE THREAD

**Issue by arizzuto** [lmcinnes/hdbscan#36]
Your HDBSCAN implementation is impressive and can be vary useful in various fields. The implementation mentions lots of different metrics, but no way to specify keywords for certain metrics. This might be an issue with me not being a python expert. If that's the case, I apologize for wasting time.

**Comment by lmcinnes** [comment]
I believe this is (hopefully) a documentation issue -- we are definitely lacking in documentation; something I hope we'll get remedied in the next few months.

Ideally you should be able to use alternative metrics with something like:

    clusterer = hdbscan.HDBSCAN(metric='mahalonobis').fit(X)
    clusterer.labels_, clusterer.probabilities_

Presuming X is a suitable ndarray, each row a point in some n-space. I believe the following *should* be supported:

                  'euclidean'
                  'minkowski'
                  'manhattan'
                  'l1'
                  'chebyshev'
                  'seuclidean'
                  'mahalanobis'
                  'wminkowski'
                  'hamming'
                  'canberra'
                  'braycurtis'
                  'matching'
                  'jaccard'
                  'dice'
                  'kulsinski'
                  'rogerstanimoto'
                  'russellrao'
                  'sokalmichener'
                  'sokalsneath'
                  'haversine'

Obviously not all of those will work with all data (Hamming expects binary, Haversine requires 2D data, and so on). I admit that I haven't tested all of this particularly well, but for the most part I am using other implementations, or stealing Jake Vanderplas' implementations from his kd-tree and ball-tree implementations, so it ought to work. Please let me know if it doesn't and I'll try to look into what's going wrong.

**Comment by arizzuto** [comment]
Thanks for getting back to me so quickly. Getting something like the Mahalanobis distance running (which is unit/scale independent in each dimension of the data) opens up the potential uses for this algorithm to lots of physical processes.  I'm basically running though one of your python notebooks, but switching in the Mahalanobis metric in place of the minkowski metric. I've added what I've done so far in the code quote below.
The issue seems to be there's no way to pass the Mahalanobis covariance matrix to the metric code. i.e. the p number for the minkowski metric doesn't get passed to the mahalanobis metric function.

`import numpy as np
import hdbscan
import matplotlib.pyplot as plt
import pdb
import seaborn as sns
import sklearn.datasets as data
moons, _ = data.make_moons(n_samples=50, noise=0.05)
blobs, _   = data.make_blobs(n_samples=50, centers=[(-0.75,2.25), (1.0, 2.0)], cluster_std=0.25)
test_data = np.vstack([moons, blobs])
##The mahalanobis covar matrix
ss = np.array([[1.0,0],[0.0,1.0]])
##!!Crashes on this line below!!
clusterer = hdbscan.HDBSCAN(min_cluster_size=5,p=ss metric='mahalanobis')
##!!Crashes on above line!!
clusterer.fit(test_data)

pdb.set_trace()
print 'here'
`

p.s. not to jump the gun here or anything, but how do I credit you guys if I use this in a scientific publication?

**Comment by lmcinnes** [comment]
In principle the Mahalanobis metric should get V, or VI as the covariance matrix, p is just for minkowski. The catch may be in ensuring that those kwargs get passed all the way through -- perhaps they do not. I'll have to look into that and get back to you.

**Comment by lmcinnes** [comment]
So the short answer is I am currently not threading the appropriate kwargs all the way through everything, hence the breakage. I'll try and get that sorted as soon as I can, and then hopefully you can pull from master here and have something that works for you.

**Comment by arizzuto** [comment]
Thanks Imcinnes! Very appreciated

**Comment by lmcinnes** [comment]
I've just committed what should be the relevant changes to master. If you can clone from here, build, and see if that works that would be great!

**Comment by lmcinnes** [comment]
As for citation -- should it ever come to that -- I would suggest you be sure to cite the Campello, Moulavi and Sander Paper(s) we reference in the README and code, and ultimately beyond that you can just cite the github repo (although I'm not sure of exactly what the right approach to that is these days). Thanks for taking an interest in our code -- we hope it proves useful to you in your research!

**Comment by arizzuto** [comment]
Thanks again for the fast work on this. I ran it and got numbers out that look reasonable, so I guess it's all working! If output from your code ends up in a paper I will be sure to let you know before publication.
